docs/releases/3.2.txt

```diff
@@ -358,6 +358,11 @@ Miscellaneous
   :attr:`~django.db.models.Model._default_manager` to check that related
   instances exist.
 
+* The admin now installs a catch all URL pattern that redirects unauthenticated
+  users to the login rather than returning a 404 response. This protects
+  against a model enumeration security issue. As a side effect, it prevents the
+  :setting:`APPEND_SLASH` setting from working on admin URLs.
```

Comment:
This is a pretty big usability hit. Is there a way to avoid this?

Reply:
Thanks for the review Carlton.

Unfortunately, I don't see a satisfactory way to do so without reintroducing the module enumeration security issue.

So long as `APPEND_SLASH` validates the URL-plus-slash in the URLconf, then the user could use it to discover and enumerate private URLs.

Here is how I'm seeing it:

Before this PR:

Unauthenticated user:
`/admin/myapp/mymodel/` → 302 to login (user knows model exists)
`/admin/myapp/doesnotexist/` → 404 (user knows model does not exist)

Therefore, this difference can be used to enumerate models.

With this PR:

Unauthenticated user:
`/admin/myapp/mymodel/` → 302 to login
`/admin/myapp/mymodel` → 302 to login
`/admin/myapp/doesnotexist/` → 302 to login
`/admin/myapp/doesnotexist` → 302 to login

The responses all look the same so the status code can't be used to enumerate models

With this PR, if `APPEND_SLASH` were still working:

Unauthenticated user:
`/admin/myapp/mymodel` → 301 to `/admin/myapp/mymodel/` then 302 to login (model exists)
`/admin/myapp/doesnotexist` → 302 directly to login (model does not exist)

This can once again be used to enumerate models.

---

I tried to think of a way to satisfy both needs and I was unable to come up with a good one. If you have a proposal, please let me know.

Here are ideas I experimented with before abandoning:

**When `APPEND_SLASH` is True, always redirect a 404 without a trailing slash to the URL with a trailing slash**

Problem: This has way too many false redirects. For example: `/does-not-exist.json` should not redirect to `/does-not-exist.json/`. That would be confsing.

**When `APPEND_SLASH` is True, only redirect for authenticated users**

Problem 1: authenticated users don't always have permission to view models, so those URLs should remain private as well. This would open a new model enumeration vulnerability.

Problem 2: Many URLs are intentionally public and don't require authentication. We want `APPEND_SLASH` to continue working for these users.

---

I recognize the usability issue, however, given the above IMO it is best to prevent the security issue.

---

One last thought,

We could move the admin URLs to not include the trialing slash for URLs. This would make mistyping admin URLs less likely. I imagine `APPEND_SLASH` exists mostly to fix the typo `/admin/myapp/mymodel` to `/admin/myapp/mymodel/`. I'm happy to begin working towards this goal if it means a more secure admin in the default configuration. This change in URLs would go through the normal deprecation strategy by issuing 301 redirects so users have a chance to migrate.

Reply:
> This change in URLs would go through the normal deprecation strategy by issuing 301 redirects so users have a chance to migrate.

I hate permanent redirects, imo it should be a temporary one (browser tend to cache them annoyingly good). Also it would have to be a 307 for people posting forms to the admin from normal http clients (yes even that happens :))